File: frontend/src/features/notifications/services/notificationNormalizer.js
import { NOTIFICATION_SEVERITY } from "../constants/notificationSeverety";

// Serviço que estrutura os eventos em mensagens prontas para notificação e alertas

/**
 * @param {import("../models/NotificationEvent.model").NotificationEvent} event
 */

function baseNotification({
  id,
  type,
  severity,
  title,
  message,
  reference,
  createdAt
}) {
  return {
    id,
    type,
    severity,
    title,
    message,
    reference,
    createdAt,
    read: false
  };
}

// função para calcular número de dias
function daysUntil(dataAlvo) {
  const today = new Date();
  const target = new Date(dataAlvo);

  // Zera horas para evitar erro de fuso/horário
  today.setHours(0, 0, 0, 0);
  target.setHours(0, 0, 0, 0);

  const diffMs = target - today;
  const diffDias = Math.ceil(diffMs / (1000 * 60 * 60 * 24));

  return diffDias;
}

export function normalizeNotification(event) {
  const { type, context, reference, createdAt, userId, id } = event;
  const { flags, data } = context;

  // RUPTURA
  if (type === "RUPTURE") {
    if (event.context.state === "BELOW_MINIMUM") {
      return baseNotification({
        id,
        type,
        severity: NOTIFICATION_SEVERITY.CRITICAL,
        title: "Estoque abaixo do mínimo",
        message: `O produto ${reference.id} possui apenas ${data?.currentStock} unidades em estoque.`,
        reference,
        createdAt,
        userId
      });
    }

    if (event.context.state === "ISNEAR_MINIMUM") {
      return baseNotification({
        id,
        type,
        severity: NOTIFICATION_SEVERITY.WARNING,
        title: "Estoque próximo do mínimo",
        message: `O produto ${reference.id} está próximo do limite mínimo. Com apenas ${data?.currentStock} unidades em estoque.`,
        reference,
        createdAt,
        userId
      });
    }
  }
  //VALIDADE 
  if (type === "VALIDITY"){
    if(event.context.state === "IS_EXPIRED"){
        return baseNotification({
            id,
            type,
            severity: NOTIFICATION_SEVERITY.CRITICAL,
            title: "Lote vencido",
            message: `O produto ${reference.id} está vencido há ${(-1)*daysUntil(data?.expirationDate)} dias.`,
            reference,
            createdAt,
            userId
        })
    }

    if(event.context.state === "ISNEAR_EXPIRATION"){
        return baseNotification({
            id,
            type,
            severity: NOTIFICATION_SEVERITY.WARNING,
            title: "Lote próximo do vencimento",
            message: `O produto ${reference.id} vencerá em ${daysUntil(data?.expirationDate)} dias.`,
            reference,
            createdAt,
            userId
        })
    }
  }
  //SUCCESS
  if (type === "VALIDITY" && event.context.state === "STOCK_UPDATED"){
        return baseNotification({
            id,
            type,
            severity: NOTIFICATION_SEVERITY.SUCCESS,
            title: "Novas movimentações adicionadas",
            message: `Nova importação ${reference.id} realizada, estoque atualizado.`,
            reference,
            createdAt,
            userId
        })
    }
    //ERROR
    if(type === "ERROR" && flags.hasInconsistency){
        return baseNotification({
            id,
            type,
            severity: NOTIFICATION_SEVERITY.CRITICAL,
            title: "Erro encontrado",
            message: `${data?.error}`,
            reference,
            createdAt,
            userId
        })
    }
    //SUGGESTION
    if(type === "SUGGESTION" && event.context.state === "SUGGEST_REPLENISHMENT"){
        return baseNotification({
            id,
            type,
            severity: NOTIFICATION_SEVERITY.INFO,
            title: "Sugestão de demanda",
            message: `Melhor momento para repor o produto ${reference.id}, veja porque...`,
            reference,
            createdAt,
            userId
        })
    }

    return null;
}

File: backend/backend/services/notification_normalizer.py
import json
from typing import Optional, Dict, Any
from backend.database.schemas import NotificationCreate
from backend.database.schemas import (
    NotificationSeverity,
    NotificationEventType
)
from datetime import datetime, date

# função para calcular número de dias
def days_until(data_alvo: str) -> int:
    """
    data_alvo deve estar no formato 'YYYY-MM-DD'
    """
    today = date.today()
    target = datetime.strptime(data_alvo, "%Y-%m-%d").date()

    diff_dias = (target - today).days
    return diff_dias


def normalize_event(evento: dict) -> Optional[NotificationCreate]:
    event_type = evento.get("type")

    context = evento.get("context")
    if isinstance(context, str):
        context = json.loads(context)
    context = context or {}

    reference = evento.get("reference")
    if isinstance(reference, str):
        reference = json.loads(reference)
    reference = reference or {}

    state = context.get("state")
    data = context.get("data", {})

    base = {
        "type": event_type,
        "reference": reference,
        "event_id": evento.get("id"),
        "user_id": evento.get("user_id"),
    }

    nome = reference.get("nome")
    nome_texto = nome if nome else "o item"

    # ======================
    # RUPTURE
    # ======================
    if event_type == "RUPTURE":

        if state == "BELOW_MINIMUM":
            return NotificationCreate(
                **base,
                severity=NotificationSeverity.CRITICAL,
                title="Estoque abaixo do mínimo",
                message=(
                    f"O produto {nome_texto} possui apenas "
                    f"{data.get('currentStock')} unidades em estoque."
                ),
            )

        if state == "NEAR_MINIMUM":
            return NotificationCreate(
                **base,
                severity=NotificationSeverity.WARNING,
                title="Estoque próximo do mínimo",
                message=(
                    f"O produto {nome_texto} está próximo do limite mínimo "
                    f"de {data.get('minimumStock')} unidades."
                ),
            )

    # ======================
    # IMPORTAÇÃO
    # ======================

    if state == "IMPORT_SUCCESS":
        file_name = data.get("file_name", "arquivo")
        inserted = data.get("inserted", 0)
        rejected = data.get("rejected", 0)

        return NotificationCreate(
            **base,
            severity=NotificationSeverity.SUCCESS,
            title="Importação concluída",
            message=(
                f"Arquivo {file_name} processado. "
                f"{inserted} registros importados e "
                f"{rejected} rejeitados."
            ),
        )

    if state == "IMPORT_PARTIAL":

        file_name = data.get("file_name", "arquivo")
        inserted = data.get("inserted", 0)
        rejected = data.get("rejected", 0)

        return NotificationCreate(
            **base,
            severity=NotificationSeverity.WARNING,
            title="Importação parcialmente concluída",
            message=(
                f"O arquivo {file_name} foi processado com inconsistências. "
                f"{inserted} registros importados e "
                f"{rejected} rejeitados."
            ),
        )

    if state == "IMPORT_ERROR":

        file_name = data.get("file_name", "arquivo")
        rejected = data.get("rejected", 0)

        return NotificationCreate(
            **base,
            severity=NotificationSeverity.CRITICAL,
            title="Falha na importação",
            message=(
                f"A importação do arquivo {file_name} falhou. "
                f"{rejected} registros foram rejeitados."
            ),
        )

    # ======================
    # VALIDITY
    # ======================
    if event_type == "VALIDITY":
        exp_date = data.get("expirationDate")
        if not exp_date:
            return None

        dias = days_until(exp_date)

        if state == "EXPIRED":
            return NotificationCreate(
                **base,
                severity=NotificationSeverity.CRITICAL,
                title="Produto vencido",
                message=f"O produto {nome_texto} está vencido há {abs(dias)} dias."
            )

        if state == "NEAR_EXPIRATION":
            return NotificationCreate(
                **base,
                severity=NotificationSeverity.WARNING,
                title="Produto próximo do vencimento",
                message=f"O produto {nome_texto} vencerá em {dias} dias."
            )

    # ======================
    # ERROR
    # ======================
    if event_type == "ERROR":
        return NotificationCreate(
            **base,
            severity=NotificationSeverity.CRITICAL,
            title="Erro encontrado",
            message=data.get("error", "Erro desconhecido.")
        )

    # ======================
    # SUGGESTION
    # ======================
    if event_type == "SUGGESTION" and state == "SUGGEST_REPLENISHMENT":
        return NotificationCreate(
            **base,
            severity=NotificationSeverity.INFO,
            title="Sugestão de reposição",
            message=(
                f"Pode ser um bom momento para repor o produto {nome_texto}. "
                f"Verifique o painel de estoque."
            )
        )

    return None